doc/heps/hep2.md

```diff
@@ -0,0 +1,366 @@
+
+  - The warning message:
+    - **Must** indicate that the feature is deprecated and is going to be removed in a future version.
+    - If applicable, **must** suggest replacement APIs.
+    - Can indicate in which version the feature was deprecated.
+    - Can indicate before which version the feature is going to be remove, however, maintainers **must** ensure that the deprecation duration (see below) is going to be respected, and that no obsolete warning will be released (e.g. feature announced to be removed in version 1.1 but is still present in 1.1).
+
+  - The warning type:
+  
+    - **Must** be a subclass of `DeprecationWarning`, `FutureWarning`, or `PendingDeprecationWarning`.
+    - **Must** be defined based on this approach (see the Appendix for more details):
+
+      1. Start by emitting a `DeprecationWarning`, to inform Library Developers, and some Data Analysts.
+      2. After 12 months, upgrade to a `FutureWarning`, to inform all users.
+
+      Exceptions are allowed but should in practice be very rare:
+
+      - Start by emitting a `PendingDeprecationWarning` for deprecations that might be reverted or are expected to be extremely noisy.
+      - Don't upgrade to a `FutureWarning` when the feature removed is meant for Library Developers only.
+      - Emit directly a `FutureWarning` when the feature removed is meant for Data Analysts only.
+
+  - `stacklevel` **must** be set so the warning appears in the correct place. Projects can implement a utiliy like the private Pandas' function [`find_stack_level`](https://github.com/pandas-dev/pandas/blob/b8a4691647a8850d681409c5dd35a12726cd94a1/pandas/util/_exceptions.py#L34) to automatically infer the right `stacklevel` number.
+
+- Documentation: all usage of the deprecated feature **must** be removed from the documentation, except from the section that serves as the reference API. Special cases (e.g. major API change) will need extra documentation.
+- Release notes: the release notes **must** list the deprecated feature.
+- Repository: all the open issues an PRs associated with the deprecated feature **must** be closed.
+
+When the deprecation involves the user having to change their code in a significant way (typically in a major release), maintainers should consider writing a migration guide.
+
+When a feature has been deprecated, its warning and documentation **must** stay in place for some time long enough to let most users find out about it. For example, Panel users who lock the dependencies of their application should be given sufficient time between the deprecation of a feature and its removal so as not to miss the deprecation warnings and be left with broken code, once they attempt to upgrade their version of Panel to the latest. A deprecated feature **can only be removed 18 months after users have been informed of its deprecation**. This period, introduced by this HEP, factors in the facts that:
```

Comment:
I don't think prescribing a fixed time period really works for us. In theory semantic versioning solves this problem by setting clear expectation that a major release can break things at will, even without any announcement whatsoever. My main point here is that we can attempt to provide some predictability around the API with this but in practice there are major changes that can happen with relatively little lead time, e.g. think of the Bokeh layout refactor and its effect on Panel 1.x. Until about a month or two before the release it was still quite unclear what the actual effect of the change would be, so despite the fact that API changes are telegraphed that actually guarantees very little about compatibility between major versions.

Reply:
> I don't think prescribing a fixed time period really works for us. 

The HEP doesn't prescribe a fixed time period, instead, it sets a minimum time period.

> In theory semantic versioning solves this problem by setting clear expectation that a major release can break things at will, even without any announcement whatsoever. If people care about that they can appropriately pin major releases.

It doesn't work quite like that in practice right 🙃 

> My main point here is that we can attempt to provide some predictability around the API with this but in practice there are major changes that can happen with relatively little lead time, e.g. think of the Bokeh layout refactor and its effect on Panel 1.x. Until about a month or two before the release it was still quite unclear what the actual effect of the change would be, so despite the fact that API changes are telegraphed that actually guarantees very little about compatibility between major versions.

Yes, my main point is to add consistency and indeed predictability. I agree that major versions are a special case and I can update the HEP accordingly, but I would still recommend giving people enough time when possible. And yes, there will always be exceptions, but I hope there will be just a few of them. The example you give is typically the sort of thing that I hope would not have happened in such a short time, for example, that didn't leave so much time to plan how to migrate CSS.